### PR TITLE
WIP: Wrapping up Host plans

### PR DESCRIPTION
**Pull request by kewitz**
Contributes to: https://github.com/opencollective/opencollective-api/pull/3076

### CODE REVIEW

File: server/models/Collective.js [hunk 1]
```diff
@@ -868,6 +868,23 @@ export default function(Sequelize, DataTypes) {
...
 
+  /**
+   * If the collective is a host, it needs to remove existing hosted collectives before
+   * deactivating it as a host.
+   */
+  Collective.prototype.deactivateAsHost = async function() {
+    const hostedCollectivesCount = await this.getHostedCollectivesCount();
+    if (hostedCollectivesCount >= 1) {
+      throw new Error(
+        `You cannot deactivate your collective while still hosting ${hostedCollectivesCount} other collectives. Please contact support.`,
```
Comment (znarf) [inline server/models/Collective.js]: The copy is unclear, it's not deactivating the collective.

Please contact support: support@opencollective.com.

File: server/graphql/v1/mutations/orders.js [hunk 2]
```diff
@@ -641,65 +642,61 @@ export async function completePledge(remoteUser, order) {
...
+    UserId: order.CreatedByUserId,
+    data: {
+      subscription: order.Subscription,
+      collective: order.collective.minimal,
+      user: remoteUser.minimal,
+      fromCollective: order.fromCollective.minimal,
+    },
+  });
+
+  if (order.Tier && order.collective.slug === PLANS_COLLECTIVE_SLUG) {
```
Comment (znarf) [inline server/graphql/v1/mutations/orders.js]: Maybe add a code comment explaining why we're doing that.

File: server/graphql/v1/mutations/orders.js [hunk 3]
```diff
@@ -641,65 +642,61 @@ export async function completePledge(remoteUser, order) {
...
+    const plan = findKey(plans, { slug: order.Tier.slug });
+    if (plan && Boolean(order.fromCollective.plan)) {
+      await order.fromCollective.update({ plan: null });
```
Comment (znarf) [inline server/graphql/v1/mutations/orders.js]: We have to think about people upgrading from a plan to another, the best way will be to subscribe to the new plan first, then cancel the old one.

Maybe we should not do that synchronously in the end. Daily cron job and alert us?

File: server/graphql/v1/mutations/orders.js (hunk 2 above)
Comment (kewitz) [reply server/graphql/v1/mutations/orders.js]: Deprecate this in favor of a worker that will properly process downgrades once a month.

File: server/lib/plans.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import { findKey, get } from 'lodash';
+
+import plans, { PLANS_COLLECTIVE_SLUG } from '../constants/plans';
+
+const isSubscribeOrUpgrade = (newPlan: string, oldPlan?: string | null): boolean => {
+  return !oldPlan ? true : get(plans, `${newPlan}.tier`) > get(plans, `${oldPlan}.tier`);
```
Comment (znarf) [inline server/lib/plans.ts]: Maybe `level` would be better so we don't make mistake about what `tier` is.

**Pull request merged**